integrations/github/save.go

```diff
@@ -0,0 +1,87 @@
+package github
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"go.uber.org/zap"
+
+	"go.autokitteh.dev/autokitteh/integrations/github/internal/vars"
+	"go.autokitteh.dev/autokitteh/sdk/sdkintegrations"
+	"go.autokitteh.dev/autokitteh/sdk/sdktypes"
+)
+
+// handleSave acts as a passthrough for the OAuth connection mode,
+// to save OAuth details (custom client ID & secrets).
+func (h handler) handleSave(w http.ResponseWriter, r *http.Request) {
+	c, l := sdkintegrations.NewConnectionInit(h.logger, w, r, desc)
+
+	// Check "Content-Type" header.
+	contentType := r.Header.Get(headerContentType)
+	if !strings.HasPrefix(contentType, contentTypeForm) {
+		c.AbortBadRequest("unexpected content type")
+		return
+	}
+
+	// Read and parse POST request body.
+	if err := r.ParseForm(); err != nil {
+		l.Warn("Failed to parse incoming HTTP request", zap.Error(err))
+		c.AbortBadRequest("failed to parse form data")
+		return
+	}
+
+	clientID := r.FormValue("client_id")
```

Comment:
Instead of all these variables, maybe have a struct and pass it along?

```go
type AuthValues struct {
    clientID string
    clientSected string
    // ...
}

func formAuth(vals url.Values) AuthValues {
     return AuthValues {
         clientID: vals.Get("client_id"),
         clientSecret: vals.Get("client_secret"),
        // ...
}
```

You can also add a `Validate` method to this struct.

Reply:
Alternatively, why read these values here pass them to `saveClientIDAndSecret`, instead of reading them inside it?
Personally, I wouldn't try to validate them because right after you save them you start an OAuth flow - it anything is wrong the flow will simply fail, which is the best kind of black-box verification.